File: tests/benchmark/matmul/bench_matmul.cpp
#include <vector>
#include "benchmark/common/benchmark_framework.h"
#include "benchmark/common/timer.h"
#include "origin.h"
#include "origin/operators/math/mat_mul.h"

using namespace origin;
using namespace origin::benchmark;
namespace F = origin::functional;

/**
 * @brief MatMul算子基准测试类
 */
class MatMulBenchmark : public BenchmarkFramework
{
public:
    double run_benchmark(const BenchmarkConfig &config) override
    {
        // MatMul算子需要两个shape
        if (config.shapes.size() != 2)
        {
            THROW_RUNTIME_ERROR("MatMul benchmark requires exactly 2 shapes, got {}", config.shapes.size());
        }

        const Shape &shape_a = config.shapes[0];
        const Shape &shape_b = config.shapes[1];

        // 创建输入张量
        size_t numel_a = shape_a.elements();
        size_t numel_b = shape_b.elements();
        std::vector<float> data_a(numel_a, 1.0f);
        std::vector<float> data_b(numel_b, 2.0f);

        auto x0 = Tensor(data_a, shape_a, origin::dtype(config.dtype).device(config.device));
        auto x1 = Tensor(data_b, shape_b, origin::dtype(config.dtype).device(config.device));

        // 检查是否启用了就地操作
        if (config.inplace)
        {
            loge("Error: MatMul operator does not support inplace operations in OriginDL");
            THROW_RUNTIME_ERROR("MatMul operator does not support inplace operations in OriginDL");
        }

        // 预热
        for (int i = 0; i < config.warmup_cnt; ++i)
        {
            auto result = F::mat_mul(x0, x1);
        }
        // 预热结束后同步，确保预热完成
        if (config.device.type() == DeviceType::kCUDA)
        {
            cuda::synchronize();
        }

        // 正式测试
        Timer timer;
        timer.start();

        for (int i = 0; i < config.repeat_cnt; ++i)
        {
            auto result = F::mat_mul(x0, x1);
        }
        // 正式测试结束后同步，确保所有计算完成
        if (config.device.type() == DeviceType::kCUDA)
        {
            cuda::synchronize();
        }

        double total_time_us = timer.elapsed_us();
        return total_time_us / config.repeat_cnt;
    }

    std::vector<std::vector<Shape>> get_default_shapes() const override
    {
        // 对于matmul，每个元素是包含两个Shape的vector
        return {
            {Shape({1, 1}), Shape({1, 1})},                  // {1, 1} x {1, 1} -> {1, 1}
            {Shape({10, 10}), Shape({10, 10})},              // {10, 10} x {10, 10} -> {10, 10}
            {Shape({100, 100}), Shape({100, 100})},          // {100, 100} x {100, 100} -> {100, 100}
            {Shape({1000, 1000}), Shape({1000, 1000})},      // {1000, 1000} x {1000, 1000} -> {1000, 1000}
            {Shape({10000, 10000}), Shape({10000, 10000})},  // {10000, 10000} x {10000, 10000} -> {10000, 10000}
        };
    }

    size_t get_required_shapes_count() const override { return 2; }

    void validate_shapes(const std::vector<Shape> &shapes) const override
    {
        if (shapes.size() != 2)
        {
            THROW_RUNTIME_ERROR("MatMul requires exactly 2 shapes, got {}", shapes.size());
        }

        const Shape &shape_a = shapes[0];
        const Shape &shape_b = shapes[1];

        // 验证至少是2维
        if (shape_a.ndims() < 2 || shape_b.ndims() < 2)
        {
            THROW_RUNTIME_ERROR("MatMul requires at least 2D shapes, got shapes with {} and {} dimensions",
                                shape_a.ndims(), shape_b.ndims());
        }

        // 验证矩阵乘法维度兼容性
        if (shape_a.ndims() == 2 && shape_b.ndims() == 2)
        {
            if (shape_a[1] != shape_b[0])
            {
                THROW_RUNTIME_ERROR("MatMul dimension mismatch: {} x {} (shape_a[1]={} != shape_b[0]={})",
                                    shape_a.to_string(), shape_b.to_string(), shape_a[1], shape_b[0]);
            }
        }
        else if (shape_a.ndims() == 3 && shape_b.ndims() == 2)
        {
            // 批量矩阵乘法：{batch, m, k} x {k, n} -> {batch, m, n}
            if (shape_a[2] != shape_b[0])
            {
                THROW_RUNTIME_ERROR("MatMul dimension mismatch: {} x {} (shape_a[2]={} != shape_b[0]={})",
                                    shape_a.to_string(), shape_b.to_string(), shape_a[2], shape_b[0]);
            }
        }
        else if (shape_a.ndims() == 3 && shape_b.ndims() == 3)
        {
            // 批量矩阵乘法：{batch, m, k} x {batch, k, n} -> {batch, m, n}
            if (shape_a[0] != shape_b[0] || shape_a[2] != shape_b[1])
            {
                THROW_RUNTIME_ERROR("MatMul dimension mismatch: {} x {} (batch or k dimension mismatch)",
                                    shape_a.to_string(), shape_b.to_string());
            }
        }
    }

    std::string get_operator_name() const override { return "MatMul"; }

    std::string get_additional_help() const override
    {
        return "                            Note: For matmul, use two shapes separated by ':'\n"
               "                            e.g., \"100,200:200,50\" for {100,200} x {200,50} -> {100,50}";
    }
};

int main(int argc, char *argv[])
{
    MatMulBenchmark benchmark;
    return benchmark.run(argc, argv);
}

/*
./build/bin/benchmark/bench_matmul -d cuda -r 20

A100，ORIGIN_KERNEL_ALGO=0
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         7.8000              
{10, 10}:{10, 10}                       20              cuda:0          float32         7.8500              
{100, 100}:{100, 100}                   20              cuda:0          float32         8.0000              
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         638.5500            
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         1559682.3000 

A100，ORIGIN_KERNEL_ALGO=1
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         7.6500              
{10, 10}:{10, 10}                       20              cuda:0          float32         7.7000              
{100, 100}:{100, 100}                   20              cuda:0          float32         7.9000              
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         604.7500            
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         983119.1000 

A100，ORIGIN_KERNEL_ALGO=2
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         7.7500              
{10, 10}:{10, 10}                       20              cuda:0          float32         7.6000              
{100, 100}:{100, 100}                   20              cuda:0          float32         7.7000              
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         618.1500            
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         981852.6500

A100，ORIGIN_KERNEL_ALGO=3
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         8.2000              
{10, 10}:{10, 10}                       20              cuda:0          float32         8.0000              
{100, 100}:{100, 100}                   20              cuda:0          float32         14.2000             
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         1206.1500           
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         1304046.5500

A100，ORIGIN_KERNEL_ALGO=4

A100，ORIGIN_KERNEL_ALGO=5
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         7.6000              
{10, 10}:{10, 10}                       20              cuda:0          float32         7.5000              
{100, 100}:{100, 100}                   20              cuda:0          float32         20.8000             
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         1229.9500           
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         1368403.3000 

A100，ORIGIN_KERNEL_ALGO=6
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         7.4000              
{10, 10}:{10, 10}                       20              cuda:0          float32         7.4500              
{100, 100}:{100, 100}                   20              cuda:0          float32         20.6000             
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         1233.5000           
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         1353519.8500 

A100，ORIGIN_KERNEL_ALGO=7
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         9.1500              
{10, 10}:{10, 10}                       20              cuda:0          float32         10.5000             
{100, 100}:{100, 100}                   20              cuda:0          float32         61.0000             
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         480.0000            
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         174940.3500

A100，ORIGIN_KERNEL_ALGO=8
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         7.9000              
{10, 10}:{10, 10}                       20              cuda:0          float32         7.8000              
{100, 100}:{100, 100}                   20              cuda:0          float32         11.4500             
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         296.3500            
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         402449.2500

A100，ORIGIN_KERNEL_ALGO=9
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         8.4000              
{10, 10}:{10, 10}                       20              cuda:0          float32         8.0000              
{100, 100}:{100, 100}                   20              cuda:0          float32         21.4500             
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         128.1500            
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         181679.7000 

A100，ORIGIN_KERNEL_ALGO=6666
shape                                   repeat          device          dtype           origindl_time_us    
{1, 1}:{1, 1}                           20              cuda:0          float32         7.9500              
{10, 10}:{10, 10}                       20              cuda:0          float32         8.2000              
{100, 100}:{100, 100}                   20              cuda:0          float32         8.2500              
{1000, 1000}:{1000, 1000}               20              cuda:0          float32         125.6500            
{10000, 10000}:{10000, 10000}           20              cuda:0          float32         179499.2500


export ORIGIN_KERNEL_ALGO=6666
python3 run_benchmark.py -f matmul -d cuda -r 20

=================================================================================================
Matmul Operator Performance Comparison
=================================================================================================
Shape                         Repeat   Device   Dtype     OriginDL(us)    PyTorch(us)     Speedup
-------------------------------------------------------------------------------------------------
{1,1}:{1,1}                   20       cuda:0   float32   8.2000          23.4699         2.8622 
{10,10}:{10,10}               20       cuda:0   float32   8.2000          22.2084         2.7083 
{100,100}:{100,100}           20       cuda:0   float32   8.4000          21.8670         2.6032 
{1000,1000}:{1000,1000}       20       cuda:0   float32   125.3500        142.9909        1.1407 
{10000,10000}:{10000,10000}   20       cuda:0   float32   189848.7500     181272.7210     0.9548 
*/

File: tests/benchmark/matmul/bench_matmul.py
"""
MatMul算子性能测试
测试PyTorch的matmul操作性能
"""

import sys
from pathlib import Path

# 添加common目录到路径
common_dir = Path(__file__).parent.parent / 'common'
if str(common_dir) not in sys.path:
    sys.path.insert(0, str(common_dir))

from typing import List, Tuple
import torch
from benchmark_framework import BenchmarkFramework, BenchmarkConfig
from timer import Timer


class MatMulBenchmark(BenchmarkFramework):
    """MatMul算子基准测试类"""
    
    def get_required_shapes_count(self) -> int:
        """MatMul算子需要两个shape"""
        return 2
    
    def run_benchmark(self, config: BenchmarkConfig) -> float:
        """执行PyTorch的matmul操作性能测试"""
        # MatMul算子需要两个shape
        if len(config.shapes) != 2:
            raise ValueError(f"MatMul benchmark requires exactly 2 shapes, got {len(config.shapes)}")
        
        shape_a = config.shapes[0]
        shape_b = config.shapes[1]
        
        # 解析设备
        device = self._parse_device(config.device)
        dtype = torch.float32 if config.dtype == 'float32' else torch.float64
        
        # 创建输入张量
        x0 = torch.ones(shape_a, dtype=dtype, device=device)
        x1 = torch.full(shape_b, 2.0, dtype=dtype, device=device)
        
        # 预热
        for _ in range(config.warmup_cnt):
            result = torch.matmul(x0, x1)
            # 确保计算完成（对于CUDA，需要同步）
            if device.type == 'cuda':
                torch.cuda.synchronize()
        
        # 正式测试
        timer = Timer()
        timer.start()
        
        for _ in range(config.repeat_cnt):
            result = torch.matmul(x0, x1)
            # 确保计算完成
            if device.type == 'cuda':
                torch.cuda.synchronize()
        
        total_time_us = timer.elapsed_us()
        return total_time_us / config.repeat_cnt
    
    def get_default_shapes(self):
        """获取默认测试形状"""
        # 对于matmul，每个元素是包含两个shape的列表
        return [
            [(1, 1), (1, 1)],
            [(10, 10), (10, 10)],
            [(100, 100), (100, 100)],
            [(1000, 1000), (1000, 1000)],
            [(10000, 10000), (10000, 10000)],
        ]
    
    def validate_shapes(self, shapes: List[Tuple[int, ...]]) -> None:
        """验证形状是否有效"""
        if len(shapes) != 2:
            raise ValueError(f"MatMul requires exactly 2 shapes, got {len(shapes)}")
        
        shape_a = shapes[0]
        shape_b = shapes[1]
        
        # 验证至少是2维
        if len(shape_a) < 2 or len(shape_b) < 2:
            raise ValueError(
                f"MatMul requires at least 2D shapes, got shapes with {len(shape_a)} and {len(shape_b)} dimensions"
            )
        
        # 验证矩阵乘法维度兼容性
        if len(shape_a) == 2 and len(shape_b) == 2:
            # 标准矩阵乘法：{m, k} x {k, n} -> {m, n}
            if shape_a[1] != shape_b[0]:
                raise ValueError(
                    f"MatMul dimension mismatch: {shape_a} x {shape_b} "
                    f"(shape_a[1]={shape_a[1]} != shape_b[0]={shape_b[0]})"
                )
        elif len(shape_a) == 3 and len(shape_b) == 2:
            # 批量矩阵乘法：{batch, m, k} x {k, n} -> {batch, m, n}
            if shape_a[2] != shape_b[0]:
                raise ValueError(
                    f"MatMul dimension mismatch: {shape_a} x {shape_b} "
                    f"(shape_a[2]={shape_a[2]} != shape_b[0]={shape_b[0]})"
                )
        elif len(shape_a) == 3 and len(shape_b) == 3:
            # 批量矩阵乘法：{batch, m, k} x {batch, k, n} -> {batch, m, n}
            if shape_a[0] != shape_b[0] or shape_a[2] != shape_b[1]:
                raise ValueError(
                    f"MatMul dimension mismatch: {shape_a} x {shape_b} "
                    f"(batch or k dimension mismatch)"
                )


def benchmark_matmul_comparison(device_filter=None,
                                shape_filter=None,
                                warmup_cnt=5,
                                repeat_cnt=100,
                                verbose=False):
    """
    运行matmul算子的PyTorch性能测试
    
    Args:
        device_filter: 设备过滤，'cpu'或'cuda'
        shape_filter: shape过滤，例如 '100,200:200,50'
        warmup_cnt: 预热次数
        repeat_cnt: 重复次数
        verbose: 是否输出详细信息
    
    Returns:
        PyTorch测试结果列表，每个元素为:
        {'shape': str, 'device': str, 'dtype': str, 'time_us': float}
    """
    benchmark = MatMulBenchmark()
    pytorch_results = benchmark.run(
        device_filter=device_filter,
        shape_filter=shape_filter,
        warmup_cnt=warmup_cnt,
        repeat_cnt=repeat_cnt,
        verbose=verbose
    )
    
    return pytorch_results


if __name__ == "__main__":
    """命令行入口，与C++版本保持一致，可以直接运行此文件进行测试"""
    benchmark = MatMulBenchmark()
    sys.exit(benchmark.run_from_command_line())
